### ISSUE TITLE
K8SSAND-98 ⁃ Speed up restores by downloading backups in parallel

### ISSUE THREAD

**Issue by jsanda** [k8ssandra/k8ssandra#235]
**Is your feature request related to a problem? Please describe.**
Restoring a backup is inherently slow right now. Downloading the backup is currently in an O(N) operation. Let me provide a brief summary of how things work.

k8ssandra configures the CassandraDatacenter an application container named `medusa` that runs a gRPC service to perform backups. It also configures the CassandraDatacenter with an initContainer named `medusa-restore`. When a CassandraRestore object is created medusa-operator makes some updates to the CassandraDatacenter to be restored. In particular it sets a couple environment variables on the `medusa-restore` container that tell it which backup to restore. Updating the CassandraDatacenter in effect triggers a rolling restart of the C* cluster. 

Since initContainers run before the main, application containers, `medusa-restore` will run before the `cassandra` container is started. It downloads the backup and updates the data directory. This all works perfectly fine. The problem is that `medusa-restore` runs serially. If we were to simply run it across the entire cluster in parallel in order to speed up downloads, then this would necessarily requiring bringing the entire cluster down. We should be able to speed up downloads without having to incur downtime.

**Describe the solution you'd like**
We add a method to the Medusa gRPC service for downloading backups. With this in place medusa-operator can easily request downloads to happen in parallel while the C* cluster is still up and running. medusa-operator then updates the CassandraDatacenter as previously described to trigger the cluster rolling restart. The `medusa-restore` container is now only responsible for updating the data directory. Downloading the backup now becomes an O(1) operation.

We probably want to add another volume for downloading the backup. Both the `medusa` and `medusa-restore` containers will need to mount this volume.

This will require changes in both Medusa and in medusa-operator.

Due to https://github.com/k8ssandra/k8ssandra/issues/70 restoring to a new cluster currently does not work. In order to support parallel downloads when restoring to a new cluster, we may need to make some changes in cass-operator. We could wait for all pods to become ready, make the request to the gRPC service to do the downloads and then do the rolling restart. This should work but it involves an unnecessary rolling restart. I need to think more about how we can handle this scenario but I am confident it is doable.

Note that the purpose of this ticket is to discuss design details. I expect the actual work to be spread out across multiple tickets.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-98) by [Unito](https://www.unito.io)
┆friendlyId: K8SSAND-98
┆priority: Medium

**Comment by jsanda** [comment]
In the description I talked having the `medusa` container download the backup files so that the downloads can be performed in parallel. This seems reasonable for an in-place restore, but it won't work for a remote-restore. The `medusa-restore` initContainer takes care of putting files back in place as well as updating cassandra.yaml (see https://github.com/thelastpickle/cassandra-medusa/issues/303).

cass-operator configures the StatefulSets with`podManagementPolicy: Parallel` which means on creation `medusa-restore` will run in parallel. So for remote-restore, we still want to download the backup files in `medusa-restore`.

**Comment by jsanda** [comment]
Closing as won't fix
